### PR TITLE
Bug fix and improved interface for deflated solvers

### PR DESCRIPTION
**Pull request by alexstrel**
This pull request includes the following changes:
1. Fix bug for solo-precision (F)GMResDR solver
2. All parameters are exposed via command line options
3. In addition, the user is now able to decide  which external library to use in the solver and/or deflation methods. Note that Eigen library remains to be requirement, and always set as default.  It is the user's responsibility to properly link quda against a preferred library. Currently, only MAGMA is supported. This can be extended with other software options like CUSOLVE etc. 
For further details, please consult QUDA Wiki page:
https://github.com/lattice/quda/wiki/Deflated-Solvers

### CODE REVIEW

File: tests/misc.cpp [hunk 1]
```diff
@@ -1223,3 +1223,77 @@ get_quda_ver_str()
...
+  } else {
+    fprintf(stderr, "Error: invalid external library type %s\n", s);
+    exit(1);
+  }
+
+  return ret;
+}
+
+QudaExtLibType
+get_df_ext_lib_type(char* s)
```
Comment (mathiaswagner) [inline tests/misc.cpp]: seems to do the same thing as get_solve_ext_lib_type above

Comment (alexstrel) [inline tests/misc.cpp]: thanks @mathiaswagner , done.

**Comment by mathiaswagner** [comment]
Btw. Eigen is now a default requirement for QUDA, see PR #628.

**Comment by alexstrel** [comment]
@mathiaswagner since eigen is default,  a few places can be cleaned. E.g., there is no need in condition at line 318 in inv_eigcg_quda.cpp and, therefore, there must be no exception branch.

**Comment by mathiaswagner** [comment]
@alexstrel Yes, you can remove those checks.

**Comment by alexstrel** [comment]
@mathiaswagner , done.

**Comment by maddyscientist** [comment]
@alexstrel can you check that the deflation is working with MILC.  The MILC interface didn't even compile with deflation enabled after #612 (fixed with #627).

**Comment by alexstrel** [comment]
@maddyscientist yes, thanks for pointing out. Also the interface must be updated with the latest changes in the parameter set.

File: lib/deflation.cpp [hunk 2]
```diff
@@ -101,24 +105,27 @@ namespace quda {
     const int nevs_to_print = param.cur_dim;
     if(nevs_to_print == 0) errorQuda("\nIncorrect size of current deflation space. \n"); 
 
-    Complex *projm  = new Complex [param.ld*param.cur_dim];
+    std::unique_ptr<Complex[], std::default_delete<Complex[]> > projm(new Complex[param.ld*param.cur_dim]);
```
Comment (mathiaswagner) [inline lib/deflation.cpp]: std::default_delete is the default so you can omit it here for readability.

File: lib/deflation.cpp [hunk 3]
```diff
@@ -22,6 +23,9 @@ namespace quda {
   using DynamicStride = Stride<Dynamic, Dynamic>;
 #endif
 
+  static auto allocator = [=] (size_t bytes ) { return static_cast<Complex*>(mapped_malloc(bytes)); };
```
Comment (mathiaswagner) [inline lib/deflation.cpp]: This doesn't compile with clang. Seems like gcc is more permissive here. Can this be fixed?

```
error: non-local lambda expression cannot have a capture-default [clang-diagnostic-error]
  static auto allocator = [=] (size_t bytes ) { return static_cast<Complex*>(mapped_malloc(bytes)); };
                           ^
Suppressed 1 warnings (1 in non-user code).
```

File: include/quda_milc_interface.h [hunk 4]
```diff
@@ -226,60 +251,45 @@ extern "C" {
...
-		    double target_fermilab_residual,
-		    const int * const domain_overlap,
-		    const void* const fatlink,
-		    const void* const longlink,
-		    void* source,
-		    void* solution,
-		    double* const final_residual,
-		    double* const final_fermilab_residual,
-		    int* num_iters);
+
```
Comment (maddyscientist) [inline include/quda_milc_interface.h]: There should be no space between doxygen comment blocks and the function they are commenting

File: include/quda_milc_interface.h [hunk 5]
```diff
@@ -443,13 +441,7 @@ extern "C" {
...
-   * @param vec_infile input file location 
-   * @param vec_outfile output file location
-   * @param max_search_dim eigCG parameter: search space dimention
-   * @param nev eigCG parameter: how many eigenpairs to compute within one eigCG call
-   * @param deflation_grid eigCG parameter : how many eigenpairs to compute within the incremental phase (# of eigenpairs = nev*deflation_grid)
-   * @param tol_restart initCG parameter : at what tolerance value to restart initCG solver
+   * @param eig_args contains info about deflation space
    * @param rhs_idx  bookkeep current rhs
    * @param last_rhs_flag  is this the last rhs to solve?
    * @param final_residual Array of true residuals
```
Comment (maddyscientist) [inline include/quda_milc_interface.h]: There should be no space between doxygen comment blocks and the function they are commenting

File: lib/deflation.cpp [hunk 6]
```diff
@@ -101,24 +105,27 @@ namespace quda {
...
-    double *evals = new double[param.ld];
-
-    cudaHostRegister(static_cast<void *>(projm), param.ld*param.cur_dim*sizeof(Complex),  cudaHostRegisterDefault);
-    magma_Xheev(projm, param.cur_dim, param.ld, evals, sizeof(Complex));
-    cudaHostUnregister(projm);
+      memcpy(projm.get(), param.matProj, param.ld*param.cur_dim*sizeof(Complex));
+      std::unique_ptr<double[] > evals(new double[param.ld]);
 
-    delete [] evals;
+      cudaHostRegister(static_cast<void *>(projm.get()), param.ld*param.cur_dim*sizeof(Complex),  cudaHostRegisterDefault);
```
Comment (maddyscientist) [inline lib/deflation.cpp]: Is there any reason not to use QUDA's `pinned_malloc` and `pinned_free` routines here? Or perhaps even better to use `pool_pinned_malloc` and `pool_pinned_free`?

File: lib/deflation.cpp [hunk 7]
```diff
@@ -170,37 +176,37 @@ namespace quda {
...
-      magma_Xgesv(vec, param.ld, param.cur_dim, param.matProj, param.ld, sizeof(Complex));
+        magma_Xgesv(vec.get(), param.ld, param.cur_dim, param.matProj, param.ld, sizeof(Complex));
 #else
-      Map<MatrixXcd, Unaligned, DynamicStride> projm_(param.matProj, param.cur_dim, param.cur_dim, DynamicStride(param.ld, 1));
-      Map<VectorXcd, Unaligned> vec_ (vec, param.cur_dim);
-
-      VectorXcd  vec2_(param.cur_dim);
+        errorQuda("MAGMA library was not built.\n");
+#endif
+      } else {
```
Comment (maddyscientist) [inline lib/deflation.cpp]: This else statement here just catches the `extlib_type != QUDA_MAGMA_EXTLIB` case and not specifically the Eigen case.  All branch-on-extlib_type statements should be made safe to throw an error if a given `extlib_type` is not supported.

File: lib/deflation.cpp [hunk 8]
```diff
@@ -295,22 +299,26 @@ namespace quda {
...
+     if( param.eig_global.extlib_type == QUDA_MAGMA_EXTLIB ) {
 #ifdef MAGMA_LIB
-     magma_Xheev(projm, param.cur_dim, param.ld, evals, sizeof(Complex));
+       magma_Xheev(projm.get(), param.cur_dim, param.ld, evals.get(), sizeof(Complex));
 #else
-     Map<MatrixXcd, Unaligned, DynamicStride> projm_(projm, param.cur_dim, param.cur_dim, DynamicStride(param.ld, 1));
-     Map<VectorXd, Unaligned> evals_(evals, param.cur_dim);
+       errorQuda("MAGMA library was not built.\n");
+#endif
+     } else {
```
Comment (maddyscientist) [inline lib/deflation.cpp]: Another branch that needs to be made safe

File: lib/deflation.cpp [hunk 9]
```diff
@@ -101,24 +105,27 @@ namespace quda {
...
+      cudaHostUnregister(projm.get());
 #else
-    Map<MatrixXcd, Unaligned, DynamicStride> projm_(param.matProj, param.cur_dim, param.cur_dim, DynamicStride(param.ld, 1));
-    Map<MatrixXcd, Unaligned, DynamicStride> evecs_(projm, param.cur_dim, param.cur_dim, DynamicStride(param.ld, 1));
-
-    SelfAdjointEigenSolver<MatrixXcd> es_projm( projm_ );
-    evecs_.block(0, 0, param.cur_dim, param.cur_dim) = es_projm.eigenvectors();//??   
+      errorQuda("MAGMA library was not built.\n");
 #endif
+    } else {
```
Comment (maddyscientist) [inline lib/deflation.cpp]: Another branch that needs to be made safe

File: lib/inv_gmresdr_quda.cpp [hunk 10]
```diff
@@ -281,18 +281,18 @@ namespace quda {
...
 
    if(do_gels) {
-#ifdef EIGEN_GELS
-     ComputeEta<libtype::eigen_lib>(args);
-#else
-     ComputeEta<libtype::magma_lib>(args);
-#endif
+     if (  param.extlib_type == QUDA_MAGMA_EXTLIB ) { 
+       ComputeEta<libtype::magma_lib>(args);
+     } else {
```
Comment (maddyscientist) [inline lib/inv_gmresdr_quda.cpp]: Another `extlib_type` branch to be made safe

File: lib/inv_gmresdr_quda.cpp [hunk 11]
```diff
@@ -312,17 +312,17 @@ namespace quda {
...
  {
 #ifdef DEFLATEDSOLVER
    GMResDRArgs &args = *gmresdr_args;
-#ifdef USE_MAGMA
-   ComputeHarmonicRitz<libtype::magma_lib>(args);
-#else
-   ComputeHarmonicRitz<libtype::eigen_lib>(args);
-#endif
+
+   if ( param.extlib_type == QUDA_MAGMA_EXTLIB ) {
```
Comment (maddyscientist) [inline lib/inv_gmresdr_quda.cpp]: Another `extlib_type` branch to be made safe

File: include/quda.h [hunk 12]
```diff
@@ -329,6 +323,9 @@ extern "C" {
     /** The index to indeicate which chrono history we are augmenting */
     int chrono_index;
 
+    /** Which external library to use in the linear solvers (MAGMA or Eigen) */
+    QudaExtLibType extlib_type;
```
Comment (maddyscientist) [inline include/quda.h]: This parameter needs also to be put into lib/quda_fortan.F90

**Comment by maddyscientist** [comment]
@alexstrel I just reviewed the code.  I've made multiple comments where changes need to be made, though these are only minor updates needed.  We can get this merged as soon as you make these edits.  Thanks for the effort 👍

**Comment by alexstrel** [comment]
@maddyscientist thanks for comments, done.

**Comment by alexstrel** [comment]
don't know why this happened , seems to be jenkins problem
```
FAILED: cd /scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir && /scratch/jenkins/cmake-3.5.2-Linux-x86_64/bin/cmake -E make_directory /scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir//. && /scratch/jenkins/cmake-3.5.2-Linux-x86_64/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=STRICT -D generated_file:STRING=/scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir//./quda_generated_color_spinor_util.cu.o -D generated_cubin_file:STRING=/scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir//./quda_generated_color_spinor_util.cu.o.cubin.txt -P /scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir//quda_generated_color_spinor_util.cu.o.cmake
sh: 1: cannot create /tmp/tmpxft_000025cb_00000002-0: Permission denied
CMake Error at quda_generated_color_spinor_util.cu.o.cmake:268 (message):
  Error generating file
  /scratch/jenkins/.jenkins/jobs/QUDAcmake/workspace/CUDAVERSION/8.0/GPUARCH/sm_52/strict/lib/CMakeFiles/quda.dir//./quda_generated_color_spinor_util.cu.o

```

**Comment by maddyscientist** [comment]
@alexstrel There does seem to be something wrong with Jenkins at the moment, as it is occasionally failing to build (I saw the same thing on my pull #642), something for @mathiaswagner to look at.  Anyway, thanks for finishing this up, will merge now.

**Pull request merged**